### PR TITLE
parse style tags

### PR DESCRIPTION
**Pull request by jc275**
- [x]  Adds `options.styletags` to parse any on-page `<style>` tags (in addition to external stylesheets).
- [x]  Processes style tag content exactly the same way as external stylesheet text.
- [x]  Preserves relative ordering of external stylesheets and on-page style tags.
- [x]  Includes `<style>` tags that were created/modified after page load.
- [x]  Does not consider any inline `style="..."` attributes.
- [ ]  Probably fixes https://github.com/peterbe/minimalcss/issues/114

### CODE REVIEW

**Comment by jc275** [comment]
I know this is probably a massive oversimplification of what you guys were planning, but I thought it might be a good start for (what I assume is) the most likely use case.

File: tests/main.test.js [hunk 1]
```diff
@@ -73,6 +73,13 @@ test('cares only about external CSS files', async () => {
   expect(finalCss).toEqual('.external{color:red}');
 });
 
+test('cares about style tags and external CSS files', async () => {
+  const { finalCss } = await runMinimalcss('css-in-js', {
+    styletags: true
+  });
+  expect(finalCss).toEqual('.cssinjs1,.external,.inline{color:red}');
```
Comment (stereobooster) [inline tests/main.test.js]: idea is to take into account inline styles so we would know if we can safely remove `@font` and `@keyframes`, but not output as part of css, because those styles already part of html.

**Comment by stereobooster** [comment]
Hi @jc275, thanks for contribution. You can ask if scope of some tickets is not clear. I left one comment (above) which summarises idea of this feature (at least from my point of view). If you have more questions don't hesitate to ask.

**Comment by jc275** [comment]
Thanks for your comments, @stereobooster.

First, as this is an _unsolicited_ PR, I won't be offended if you think it's unconstructive and decide to close it!

Otherwise, a few clarifications if I may...

1. By 'inline styles'... are we talking about `<style>` tags, or `style="..."` attributes, or both?

2. Since the html will always have to be modified—at least, to replace the original stylesheet links—is it essential to keep inline styles where they are? I imagine that a common use case would be to merge `largely-unused-huge-boilerplate-framework.min.css` with a few custom stylesheets or style tags—which may or may not be critical-rendering-path CSS that needs to stay on the page. On that basis, would it be helpful to use an approach like [isomorphic-style-loader](https://github.com/kriasoft/isomorphic-style-loader) or [penthouse](https://github.com/pocketjoso/penthouse) (which [addyosmani/critical](https://github.com/addyosmani/critical) uses under the hood) to decide—after `minimalcss`-ification—which styles should be inlined?

(My main use case is extracting style rules from third-party websites, so I would prefer to combine inline styles with external ones. I realise that the intended purpose is for users to optimise their own websites/apps, but it seems to me that merging linked stylesheets with inline styles may sometimes be very useful for that use case, too.)

**Comment by stereobooster** [comment]
> By 'inline styles'... are we talking about <style> tags, or style="..." attributes, or both?

Both because: `<div style="animation-name: example">` and `<h1 style="font-family: GoogleFont;">`. To detect if we need `@font` and `@keyframes` we need to parse both (without flag). But if we want also inline styles to be part of the output, I would say this should be behind the flag (what you did), but I'm still not sure about use case.

**Comment by jc275** [comment]
I can't really offer any help with `@keyframes` and `@font` rules. For my specific use case, I'm only interested in other types of CSS rules.

I _can_ see a clear, common use case for merging (and perhaps splitting, based on critical render path) the linked and inline styles. However, I haven't considered `style` attributes at all, so my PR doesn't fully solve that problem.

As an aside, would it be cleaner if everything currently happening inside puppeteer's `page.evaluate(() => {...})` were simplified to:
```js
const htmlWithJavascript = await page.evaluate(() => document.documentElement.outerHTML);
const $ = cheerio.load(htmlWithJavascript);
doms.push($);
// do all the interesting stuff in the local scope, instead of the browser
$('link, style, [style]').each(utils.someHelperFunction);
```

**Comment by jc275** [comment]
Maybe it's necessary to repeat the whole process multiple times, independently minimizing the css for external stylesheets, inline styles, and both combined. Then you can do (pseudocode):
```js
// once all unused css rules have been removed...
outputRules = externalAndInlineRules.filter(rule =>
  externalRules.includes(rule) // current behaviour
  || !inlineRules.includes(rule) // for external @font and @keyframes
);
```
(I'm just thinking aloud, as it's too much for me to take on.)

**Comment by peterbe** [comment]
I haven't had a chance yet to look at the code but I skimmed it really quick. This is so exciting!

**Comment by peterbe** [comment]
You've done an excellent job but I think we need to break it down a little bit. 

@stereobooster is right. The primary objective is to not remove things like `@font` based *only* on all the external style sheets. For that, you shouldn't need to explicitly use flags/options like `--styletags`. It should just work automatically. 

I would advise that we break up your awesome code into two PRs. 

## 1. One PR just to survive these two unit tests:

```html
<html>
<link rel=stylesheet href="hasfonts.css">
<style>p { font-family: Lato; }</style>
</html>
```
...where `hasfonts.css` contains...

```css
@font-face {
  font-family: 'Lato';
  font-style: normal;
  font-weight: 400;
  src: local('Lato Regular'), local('Lato-Regular'), url(https://fonts.gstatic.com/s/lato/v14/S6uyw4BMUTPHjxAwXiWtFCfQ7A.woff2) format('woff2');
  unicode-range: U+0100-024F;
}
```

Another example of a unit test would be this:
```html
<html>
<link rel=stylesheet href="hasanimations.css">
<body>
<p style="animation:progress-bar-stripes 2s linear infinite">Hej</p>
</body>
</html>
```

...where `hasanimations.css` contains this:
```css
@keyframes progress-bar-stripes {
  from {
    background-position: 40px 0;
  }
  to {
    background-position: 0 0;
  }
}
```

To solve this, we'd need to pass two AST objects to [`postProcessOptimize`](https://github.com/peterbe/minimalcss/blob/7a43191b27a199da736261eaba56aa1a52472b66/src/run.js#L33). One AST (that gets modified) and one that we inspect. 
So, for example, instead of:
```javascript
  const activeAnimationNames = new Set(
    csstree.lexer
      .findAllFragments(ast, 'Type', 'keyframes-name')
      .map(entry => csstree.generate(entry.nodes.first()))
  );

  csstree.walk(ast, {
    visit: 'Atrule',
    enter: (node, item, list) => {
      const basename = csstree.keyword(node.name).basename;
      if (basename === 'keyframes') {
        if (!activeAnimationNames.has(csstree.generate(node.prelude))) {
          list.remove(item);
        }
      } else if (basename === 'media') {
        if (csstree.generate(node.prelude) === 'print') {
          list.remove(item);
        }
      }
    }
  });
```
perhaps something like this:
```javascript
  const activeAnimationNames = new Set(
    csstree.lexer
      .findAllFragments(astComplete, 'Type', 'keyframes-name')  // HERE! Note the use of 'astComplete'
      .map(entry => csstree.generate(entry.nodes.first()))
  );

  csstree.walk(ast, {
    visit: 'Atrule',
    enter: (node, item, list) => {
      const basename = csstree.keyword(node.name).basename;
      if (basename === 'keyframes') {
        if (!activeAnimationNames.has(csstree.generate(node.prelude))) {
          list.remove(item);
        }
      } else if (basename === 'media') {
        if (csstree.generate(node.prelude) === 'print') {
          list.remove(item);
        }
      }
    }
  });
```
Note here that in the first block we generate the set `activeAnimationNames` based on `astComplete` and in the second block we modify (in place) an object called `ast`.

To do that, we need to create two AST objects and later call:
```javascript
postProcessOptimize(allCombinedAst, allCompleteCombinedAst);
```
...[here](https://github.com/peterbe/minimalcss/blob/7a43191b27a199da736261eaba56aa1a52472b66/src/run.js#L551).

That "complete AST" (called `allCompleteCombinedAst` in the example above), needs to be created not only by downloading all external style sheets but also every single `<style>` tag **and** every `style` attribute. 

The tricky thing is that if you add the `<style>` tag's CSS payloads to the AST, that's valid CSS. But in the case of `<p style="animation:progress-bar-stripes 2s linear infinite">` the value is just `animation:progress-bar-stripes 2s linear infinite` which isn't valid CSS. So you'd have to pretend that it's actually `p{  animation:progress-bar-stripes 2s linear infinite  }`.

## 2. One PR just to extract all the `<style>` tags

I think this would be easier. But perhaps it needs to depend on code from the 1. point above. It could sit under an option/flag like `--inlinestyles` It would basically, make that AST based on something like `$('link, style').each(utils.someHelperFunction);` instead of `$('link').each(utils.someHelperFunction);` as per your example. 

I could see myself using this. That would make it possible to replace ALL `<link>` tags' content **and** all `<style>` tags' content into one single `<style>` tag.

**Comment by peterbe** [comment]
> Maybe it's necessary to repeat the whole process multiple times, independently minimizing the css for external stylesheets, inline styles, and both combined.

I think this is wrong. The point, made above, is that you might have actual connections between the CSS in the external stylesheets (defining things like keyframes) and the inline stylesheets (referring to keyframe animations by name).

**Comment by peterbe** [comment]
> As an aside, would it be cleaner if everything currently happening inside puppeteer's `page.evaluate(() => {...})` were simplified to: ...

You're referring to [this code](https://github.com/peterbe/minimalcss/blob/7a43191b27a199da736261eaba56aa1a52472b66/src/run.js#L304-L340), right? 
That's an excellent point! I suspect the code is the way it is because in the early days we didn't have `cheerio`. The job of that code is to get the whole HTML *and* the list of external stylesheet `hrefs`. That could easily be done outside of puppeteer in a utility function that uses cheerio. 

That would make for an excellent PR on the side as a nice refactoring step.

**Comment by jc275** [comment]
Wow, thanks for the comments and analysis.

Regarding `cheerio`, refactoring from `page.evaluate(...)` might produce more modular, testable code. But, on the other hand, `cheerio` has to iterate over non-standard POJO representations of each node, while chromium has native access to each `HTMLElement`. There are some advantages to the native DOM API (such as missing attributes defaulting to empty strings rather than `undefined`) which might allow us to keep the code simpler and more readable. So I'm in two minds about this.


> ## 2. One PR just to extract all the `<style>` tags
If I've understood you correctly, that's exactly what the current PR is. (I called the option `styletags` instead of `inlinestyles` because it doesn't capture any inline `style` attributes.)

**Comment by jc275** [comment]
> I would advise that we break up your awesome code into two PRs.
> ## 1. One PR just to survive these two unit tests
My awesome code doesn't address this at all :)

**Comment by jc275** [comment]
> So you'd have to pretend that it's actually `p { animation:progress-bar-stripes 2s linear infinite }`.

That could lead to problems if one `p` supersedes another.
Maybe create dummy IDs like this?
```css
#__minimalcss_inline_1__ { animation: progress-bar-stripes 2s linear infinite }
#__minimalcss_inline_2__ { animation: progress-bar-stripes 1s ease }
```

**Comment by peterbe** [comment]
> If I've understood you correctly, that's exactly what the current PR is. 

I unchecked that this PR fixes #114 in the PR description. Right?

I'll review it properly when time allows. 

We'll have to make another attempt at solving the problem of checking if a `@font-face` is mentioned in the style tags in a separate PR after this.

**Comment by jc275** [comment]
https://github.com/peterbe/minimalcss/blob/7a43191b27a199da736261eaba56aa1a52472b66/src/run.js#L304-L305
`Set` should work in chromium; I think an array is still needed because we need to preserve the ordering of stylesheets (and style tags).

**Comment by peterbe** [comment]
I've finally had time to look into this. 
First of all, you need to add this new option to `bin/minimalcss.js` so it becomes possible to do something like `--include-styletags` or something.

Secondly, I'm not sure I understand how it works. I first ran:
```
node bin/minimalcss.js -o /tmp/peterbecom.master.css https://www.peterbe.com/
```
using the `master` branch. 

Then, I checked out your branch, and before using it I manually edited to say:
```
options.styletags=true;
```
...so I can keep testing with the cli. 
Now I run:
```
node bin/minimalcss.js -o /tmp/peterbecom.css https://www.peterbe.com/
```

The one made with your branch is bigger:
```
▶ ls -l /tmp/peterbecom.*css
-rw-r--r--  1 peterbe  wheel  24876 Sep  5 20:38 /tmp/peterbecom.css
-rw-r--r--  1 peterbe  wheel  22910 Sep  5 20:38 /tmp/peterbecom.master.css
```
The difference is 1966 bytes. I *would* expect it to be 108 bytes, because that site is weird. 
It uses a bunch of `<link rel="stylesheet">` tags but it also has one `<style>` block which is never in an external style sheet. That one contains:
```
h2{margin-top:60px;padding-top:60px;border-top:1px solid #e8e9e9;font-size:2rem}div.post{margin-bottom:60px}
```
which is 108 bytes long. 

I took the two `.css` files and uploaded them here:
(using master branch) https://gist.github.com/peterbe/80332a86f9153667c4f07f6c8c0dbfb1
and
(using your branch) https://gist.github.com/peterbe/ace6c2e739b44d58675e975237076208
Before uploading them, I ran them both through `prettier` to make sense of it. 
If you look at the CSS you branch produced `div.post {...}`, for example, which was only in a style tag. 

But what's up with all the other differences?

**Comment by peterbe** [comment]
Note, https://www.peterbe.com was perhaps a weird example to use but it just happened to be the first one I tried. That site actually uses `minimalcss` to extract all the external stylesheets into a block of minimal CSS and then inserts that into the HTML with a `<style>` tag and has all the remaining external CSS loaded async and after the DOM is loaded.

**Comment by peterbe** [comment]
It worked better with `https://www.buzzfeednews.com/article/daveyalba/facebook-philippines-dutertes-drug-war`. Just a random URL. 
That page has a big fat `<style>` block which is 107,188 bytes long but after processing with your branch the whole thing is reduced down to 34,538 bytes.

**Comment by jc275** [comment]
Regarding the differences you see on peterbe.com, isn't that because of the second (much larger) style tag?

**Comment by peterbe** [comment]
> Regarding the differences you see on peterbe.com, isn't that because of the second (much larger) style tag?

Here's the thing. That site, when rendered in the browser will kinda have two copies of a bunch of things. The `<style>` block has all the critical CSS extracted from files like `https://www.peterbe.com/static/css/base.min.8c1ef15190f7.css`. Once that file has been loaded it should contain everything the `<style>` block in the HTML *contained*. 
E.g.

```html
<head>
<link rel="stylesheet" href="main.css" lazy-load=true>
<style>
A { ... }
C { ... }
E { ... }
</style>
</head>
```
and `main.css` contains:
```css
A { ... }
B { ... }
C { ... }
D { ... }
E { ... }
F { ... }
```

So, once merged it should only become:
```css
A { ... }
B { ... }
C { ... }
D { ... }
E { ... }
F { ... }
```
Nothing more. Curious.

**Comment by jc275** [comment]
After a bit of re-ordering, the diff between your two gists ends up looking like this:
```css
/*! License for minified and inlined CSS originally belongs to Semantic UI. See individual files in https://github.com/peterbe/django-peterbecom/tree/master/peterbecom/base/static/css */
@media only screen and (max-width: 767px) {
  .ui.container {
    width: auto !important;
    margin-left: 1em !important;
    margin-right: 1em !important;
  }
}
@media only screen and (min-width: 768px) and (max-width: 991px) {
  .ui.container {
    width: 723px;
    margin-left: auto !important;
    margin-right: auto !important;
  }
}
@media only screen and (min-width: 992px) and (max-width: 1199px) {
  .ui.container {
    width: 933px;
    margin-left: auto !important;
    margin-right: auto !important;
  }
}
@media only screen and (min-width: 1200px) {
  .ui.container {
    width: 1127px;
    margin-left: auto !important;
    margin-right: auto !important;
  }
}
@media only screen and (max-width: 767px) {
  .ui.segment:not(.vertical) .ui.stackable.page.grid {
    margin-left: -1rem !important;
    margin-right: -1rem !important;
  }
}
@media only screen and (max-width: 767px) {
  .ui.menu > .ui.container {
    width: 100% !important;
    margin-left: 0 !important;
    margin-right: 0 !important;
  }
}
@media only screen and (min-width: 768px) {
  .ui.menu:not(.secondary):not(.text):not(.tabular):not(.borderless)
    > .container
    > .item:not(.right):not(.borderless):first-child {
    border-left: 1px solid rgba(34, 36, 38, 0.1);
  }
}
@media (max-width: 414px) {
  h1.ui.header {
    font-size: 2.5rem;
  }
}
@media (min-width: 700px) {
  pre {
    padding: 10px;
  }
}
@media (min-width: 767px) {
  .main.menu .right.dropdown {
    display: none !important;
  }
  .main.menu .right.menu {
    display: flex;
  }
}
@media (max-width: 768px) {
  .main.menu .right.dropdown {
    display: flex !important;
  }
  .main.menu .right.menu {
    display: none !important;
  }
}
h2 {
  margin-top: 60px;
  border-top: 1px solid #e8e9e9;
}
div.post {
  margin-bottom: 60px;
}
```
Not quite sure what to make of it. Any thoughts on where all the `@media` queries are coming from?

As an aside, it doesn't look as if they're minified very effectively, as there's a lot of duplication, e.g.
```css
@media only screen and (max-width: 767px) {
  .ui.container { /* ... */ }
}
@media only screen and (max-width: 767px) {
  .ui.segment:not(.vertical) .ui.stackable.page.grid { /* ... */ }
}
@media only screen and (max-width: 767px) {
  .ui.menu > .ui.container { /* ... */ }
}
```

**Comment by peterbe** [comment]
A ha! I think I know what means. [Merging media queries is dangerous](https://github.com/css/csso/issues/162)
That's why the [`forceMediaMerge`](https://github.com/css/csso#compressast-options) is disabled by default. 

In https://github.com/peterbe/minimalcss/issues/13 I wanted it to be an option for the brave souls that dare to pass `forceMediaMerge` into `csso.compress()`. 

But it's probably best to leave that as is for now. I certainly wouldn't dare to enable it for my site. 

In conclusion. Your branch is good and healthy!! Ignore testing with www.peterbe.com.

So my only outstanding issue is that your new option needs to extended in `bin/minimalcss.js` so one can use this cool new feature on the command line.

File: src/run.js [hunk 2]
```diff
@@ -301,34 +315,76 @@ const processPage = ({
...
-          html: document.documentElement.outerHTML,
-          hrefs
-        };
+        return { html, hrefs, styles };
       });
 
       const htmlWithJavascript = evalWithJavascript.html;
       doms.push(cheerio.load(htmlWithJavascript));
+
+      if (options.styletags) {
```
Comment (peterbe) [inline src/run.js]: Can you move this up to where all the other `options` are dealt with. E.g.

```javascript
const styletags = options.styletags || false;

....

if (styletags) {
...
```

Comment (peterbe) [reply src/run.js]: Mainly for consistency and that it feels like we're done with `options` after it has been turned into variables.

File: tests/main.test.js (hunk 1 above)
Comment (peterbe) [reply tests/main.test.js]: Can you add a comment here to explain the awesomeness. 
It could be something like:
```
// The css-in-js.html fixture has an external stylesheet and a <style> 
// tag. The <style> tag contains the selector `.inline{...}`.
// This test asserts that the selectors from the inline <style> tags really
// are included in the final CSS. 

```

File: src/run.js (hunk 2 above)
Comment (jc275) [reply src/run.js]: Done.
